### PR TITLE
Presence API: improvements to channel selection at request::Subscribe

### PR DESCRIPTION
**Pull request by MOZGIII**
Follow up for #35 
After this is done, I'll work on #40 to finally complete the presence API integration!

### CODE REVIEW

**Comment by MOZGIII** [comment]
@parasyte this one's ready for a review!

**Comment by parasyte** [comment]
@MOZGIII I'm seeing some test failures on this branch, because `cargo test` is taking too long for Travis.

These tests are taking the longest:

```
     Running target/debug/deps/presence-10614ffd15733a6d

test global_here_now ... test global_here_now has been running for over 60 seconds

     Running target/debug/deps/pubsub-3de3ba578bf7dacb

test pubnub_subscribe_clone_ok ... test pubnub_subscribe_clone_ok has been running for over 60 seconds
test pubnub_subscribe_ok ... test pubnub_subscribe_ok has been running for over 60 seconds
```

**Comment by MOZGIII** [comment]
> @MOZGIII I'm seeing some test failures on this branch, because `cargo test` is taking too long for Travis.
> 
> These tests are taking the longest:
> 
> ```
>      Running target/debug/deps/presence-10614ffd15733a6d
> 
> test global_here_now ... test global_here_now has been running for over 60 seconds
> 
>      Running target/debug/deps/pubsub-3de3ba578bf7dacb
> 
> test pubnub_subscribe_clone_ok ... test pubnub_subscribe_clone_ok has been running for over 60 seconds
> test pubnub_subscribe_ok ... test pubnub_subscribe_ok has been running for over 60 seconds
> ```
> 
> There are also various lint failures.
> 

The test simply has a sleep in it. Otherwise, the PubNub network just doesn't react in time. I don't know a way to work around it. We can mock the backend, but I still want to test against the real implementation for integration purposes.

> I also see this failure on the latest available nightly (with clippy and so on): `rustc 1.44.0-nightly (f509b26a7 2020-03-18)`
> 
> ```
> test pubnub::tests::mocked_pubnub_subscribe_ok ... FAILED
> 
> failures:
> 
> ---- pubnub::tests::mocked_pubnub_subscribe_ok stdout ----
> [2020-03-24T00:09:50Z DEBUG pubnub_core::subscription::subscribe_loop_supervisor] Creating the subscribe loop
> [2020-03-24T00:09:50Z DEBUG pubnub_core::subscription::subscribe_loop_supervisor] Waiting for subscription loop ready...
> [2020-03-24T00:09:50Z DEBUG pubnub_core::subscription::subscribe_loop] Starting subscribe loop
> [2020-03-24T00:09:50Z DEBUG pubnub_core::subscription::subscribe_loop] messages: [Message { message_type: Publish, route: None, channel: Name("test_channel"), json: Object(Object { store: [("test", Short("value"), 0, 0)] }), metadata: Null, timetoken: Timetoken { t: 100, r: 12 }, client: None, subscribe_key: "test_subscribe_key", flags: 514 }]
> [2020-03-24T00:09:50Z DEBUG pubnub_core::subscription::subscribe_loop] timetoken: Timetoken { t: 150, r: 1 }
> [2020-03-24T00:09:50Z DEBUG pubnub_core::subscription::subscribe_loop] Delivering to (1, Some(1)) listeners for Channel(Name("test_channel"))...
> thread 'pubnub::tests::mocked_pubnub_subscribe_ok' panicked at 'attempted to leave type `std::mem::ManuallyDrop<std::cell::UnsafeCell<std::boxed::Box<std::boxed::Box<dyn std::ops::FnOnce() -> mock::runtime::gen::MockRuntime>>>>` uninitialized, which is invalid', /rustc/f509b26a7730d721ef87423a72b3fdf8724b4afa/src/libcore/mem/mod.rs:536:5
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 
> 
> failures:
>     pubnub::tests::mocked_pubnub_subscribe_ok
> 
> test result: FAILED. 11 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
> ```

This one is new, I'll investigate.

**Comment by MOZGIII** [comment]
I did some investigation.

The nightly things might be fluke, it's hard to say anything there, and we hardly have to look into it deeper at this point. If `beta` channel `rustc` starts to fail that same way - I'll investigate it deeper.

Both are unrelated to this PR, we can merge it.

**Comment by MOZGIII** [comment]
@parasyte mind approving this?

File: pubnub-core/src/data/channel/wildcard_spec.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+///
+/// This type represents a value suitable for use with [wildcard subscribe]
+/// feature.
+///
+/// Currently you can have up to three levels deep with your channel segment
+/// hierarchies, `a.b.c`, for example.
+///
+/// Developers MUST use .* to denote a wildcard subscription. Just having a * at
+/// the end of channel names will NOT get translated to .* and will result in a
+/// subscribe error.
```
Comment (parasyte) [inline pubnub-core/src/data/channel/wildcard_spec.rs:16]: Does it really return an error?

File: pubnub-core/src/data/channel/wildcard_spec.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+impl FromStr for WildcardSpec {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !Self::is_valid(s) {
+            return Err(());
+        }
+        Ok(Self(s.to_owned()))
+    }
+}
```
Comment (parasyte) [inline pubnub-core/src/data/channel/wildcard_spec.rs:96]: These `From` impls skip the `channel::Name` validation.

File: pubnub-hyper/src/transport/hyper/presence.rs [hunk 3]
```diff
@@ -134,8 +134,8 @@ impl TransportService<request::SetState> for Hyper {
             state,
         } = request;
 
-        let channels = EncodedChannelsList::from(channels);
-        let channel_groups = EncodedChannelsList::from(channel_groups);
+        let channels = UrlEncodedList::from(channels);
+        let channel_groups = UrlEncodedList::from(channel_groups);
```
Comment (parasyte) [inline pubnub-hyper/src/transport/hyper/presence.rs:138]: Can these reuse `pubsub::process_subscribe_to()`?

File: pubnub-core/src/subscription/message_destinations.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    fn route_destination(&self) -> Option<pubsub::SubscribeTo> {
+        match self.message.route {
+            Some(message::Route::ChannelGroup(ref val)) => {
+                Some(pubsub::SubscribeTo::ChannelGroup(val.clone()))
+            }
+            Some(message::Route::ChannelWildcard(ref val)) => {
+                Some(pubsub::SubscribeTo::ChannelWildcard(val.clone()))
+            }
```
Comment (parasyte) [inline pubnub-core/src/subscription/message_destinations.rs:27]: Cloning here is the only thing in this PR I'm not too fond of. AFAICT, these are used by the `Iterator` impl, but that could just borrow the names, and we could add an `Into<Iterator>` impl to transfer ownership of the strings where necessary.

But we can save the `clone` audit for later!

File: pubnub-hyper/src/transport/hyper/pubsub.rs [hunk 5]
```diff
@@ -85,28 +85,135 @@ impl TransportService<request::Subscribe> for Hyper {
...
+    let channels = to.iter().filter_map(|to| {
+        to.as_channel()
+            .map(AsRef::<str>::as_ref)
+            .or_else(|| to.as_channel_wildcard().map(AsRef::<str>::as_ref))
+    });
+    let channel_groups = to.iter().filter_map(pubsub::SubscribeTo::as_channel_group);
+
+    let channels = UrlEncodedList::from(channels).into_inner();
+    let channels = if channels.is_empty() {
+        "-".to_owned()
```
Comment (parasyte) [inline pubnub-hyper/src/transport/hyper/pubsub.rs:177]: Whoops, this is not right. Empty channels are technically allowed when subscribing to channel-groups. Quoting the [REST API docs](https://www.pubnub.com/docs/pubnub-rest-api-documentation#publish-subscribe-subscribe-get-v2):

> The channel name(s) you are subscribing to. Verify that channels are comprised of valid characters. You may subscribe to mulitple channels using a comma seperator. **If subscribing to no channels (only channel groups), use a comma char (,) as a placeholder.** You may subscribe to channels, channels & channel groups, or just channel groups.

Emphasis added. The "placeholder" value can be considered a two-element list with both elements being empty strings.

Comment (MOZGIII) [reply pubnub-hyper/src/transport/hyper/pubsub.rs:177]: I think I meant to use a comma there (`,`) instead if the dash. Would that make more sense?

Comment (parasyte) [reply pubnub-hyper/src/transport/hyper/pubsub.rs:177]: Yes, that will fix it! Thanks.

**Comment by MOZGIII** [comment]
Replaced with #53.
@parasyte please approve that one.

File: pubnub-hyper/src/transport/hyper/presence.rs (hunk 3 above)
Comment (MOZGIII) [reply pubnub-hyper/src/transport/hyper/presence.rs:138]: It is unclear if the logic is the same. I decided to hold on the decision to change these calls until we have more calls - and then either generalize the `SubscribeTo`, or implement more similar objects.

File: pubnub-core/src/data/channel/wildcard_spec.rs (hunk 2 above)
Comment (MOZGIII) [reply pubnub-core/src/data/channel/wildcard_spec.rs:96]: Yes, but I think this is correct, as the constraints for the names are different. I.e. they're related, but I couldn't figure out how to match them.

File: pubnub-core/src/data/channel/wildcard_spec.rs (hunk 1 above)
Comment (MOZGIII) [reply pubnub-core/src/data/channel/wildcard_spec.rs:16]: It's just copied from the docs: https://support.pubnub.com/support/solutions/articles/14000043663-how-do-i-subscribe-to-a-wildcard-channel-

**Pull request closed without merging**